File: src/core/llm/gemini.ts
import { GoogleGenAI } from '@google/genai';
import type { Message, Attachment } from '@/core/types';
import type { LLMStrategy } from './types';

const ENV_API_KEY = import.meta.env.VITE_LLM_API_KEY;

export class GeminiStrategy implements LLMStrategy {
    private client: GoogleGenAI;
    private modelName: string;

    constructor(modelName: string) {
        this.modelName = modelName;
        this.client = new GoogleGenAI({ apiKey: ENV_API_KEY });
    }

    async chat(history: Message[], newMessage: string, systemPrompt?: string, onStream?: (chunk: string) => void, attachments?: Attachment[]): Promise<string> {
        const contents = this.buildContents(history, newMessage, attachments);

        if (onStream) {
            const response = await this.client.models.generateContentStream({
                model: this.modelName,
                contents,
                config: {
                    systemInstruction: systemPrompt,
                    maxOutputTokens: 32768
                }
            });

            let fullText = '';
            for await (const chunk of response) {
                const text = chunk.text || '';
                if (text) {
                    fullText += text;
                    onStream(text);
                }
            }
            return fullText;
        }

        return this.chatSync(history, newMessage, systemPrompt, attachments);
    }

    async chatSync(history: Message[], newMessage: string, systemPrompt?: string, attachments?: Attachment[]): Promise<string> {
        const contents = this.buildContents(history, newMessage, attachments);

        const response = await this.client.models.generateContent({
            model: this.modelName,
            contents,
            config: {
                systemInstruction: systemPrompt,
                maxOutputTokens: 32768
            }
        });

        return response.text || '';
    }

    private buildContents(history: Message[], newMessage: string, newAttachments?: Attachment[]): any[] {
        const contents: any[] = [];

        // History
        history.forEach(m => {
            const parts: any[] = [{ text: m.content }];
            if (m.attachments) {
                m.attachments.forEach(a => {
                    parts.push({
                        inlineData: {
                            mimeType: a.mimeType,
                            data: a.data.split(',')[1] || a.data // Handle data:image/png;base64,...
                        }
                    });
                });
            }
            contents.push({
                role: m.role === 'assistant' ? 'model' : 'user',
                parts
            });
        });

        // New Message
        const newParts: any[] = [{ text: newMessage }];
        if (newAttachments) {
            newAttachments.forEach(a => {
                newParts.push({
                    inlineData: {
                        mimeType: a.mimeType,
                        data: a.data.split(',')[1] || a.data
                    }
                });
            });
        }
        contents.push({ role: 'user', parts: newParts });

        return contents;
    }
}

File: server/pkg/ai/sdk/gemini.go
package sdk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"

	"time"

	"github.com/soaringk/aperture/server/pkg/logger"
	"go.uber.org/zap"
	"google.golang.org/genai"
)

// GeminiProvider holds the reusable Gemini client and config
type GeminiProvider struct {
	client *genai.Client
	model  string
}

// NewGeminiProvider creates a Gemini provider with initialized client
func NewGeminiProvider(cfg Config) (*GeminiProvider, error) {
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  cfg.APIKey,
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create Gemini client: %w", err)
	}

	return &GeminiProvider{
		client: client,
		model:  cfg.Model,
	}, nil
}

// Chat handles content generation
func (p *GeminiProvider) Chat(ctx context.Context, req ChatRequest, yield func(any) error) error {
	log := logger.L()

	// Parse messages and convert to Gemini format
	var rawMessages []struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	}
	if err := json.Unmarshal(req.Messages, &rawMessages); err != nil {
		log.Error("Failed to parse messages", zap.Error(err))
		return fmt.Errorf("invalid messages format: %w", err)
	}

	var contents []*genai.Content
	for _, m := range rawMessages {
		if m.Role == "system" {
			continue // System prompt handled separately
		}
		role := m.Role
		if role == "assistant" {
			role = "model"
		}
		contents = append(contents, &genai.Content{
			Role:  role,
			Parts: []*genai.Part{genai.NewPartFromText(m.Content)},
		})
	}

	config := &genai.GenerateContentConfig{}
	if req.SystemPrompt != "" {
		config.SystemInstruction = &genai.Content{
			Parts: []*genai.Part{genai.NewPartFromText(req.SystemPrompt)},
		}
	}

	log.Debug("Sending request to Gemini",
		zap.String("model", p.model),
		zap.Int("content_count", len(contents)),
	)

	stream := p.client.Models.GenerateContentStream(ctx, p.model, contents, config)

	// Generate a stable ID for this stream
	id := fmt.Sprintf("chatcmpl-%d", time.Now().UnixNano())
	created := time.Now().Unix()
	isFirstChunk := true
	toolCallIndex := 0

	for resp, err := range stream {
		if err != nil {
			log.Error("Gemini stream error", zap.Error(err))
			return p.wrapError(err)
		}

		// Send usage from final response
		var usage *Usage
		if resp.UsageMetadata != nil {
			usage = &Usage{
				PromptTokens:     int(resp.UsageMetadata.PromptTokenCount),
				CompletionTokens: int(resp.UsageMetadata.CandidatesTokenCount),
				TotalTokens:      int(resp.UsageMetadata.TotalTokenCount),
			}
		}

		for _, cand := range resp.Candidates {
			// Map finish reason
			var finishReason *string
			if cand.FinishReason != "" {
				fr := mapGeminiFinishReason(cand.FinishReason)
				finishReason = &fr
			}

			if cand.Content != nil {
				for _, part := range cand.Content.Parts {
					// Handle text content
					if part.Text != "" {
						chunk := OpenAIChunk{
							ID:      id,
							Object:  "chat.completion.chunk",
							Created: created,
							Model:   p.model,
							Choices: []Choice{
								{
									Index: 0,
									Delta: Delta{
										Content: part.Text,
									},
									FinishReason: finishReason,
								},
							},
							Usage: usage,
						}
						// Set role on first chunk
						if isFirstChunk {
							chunk.Choices[0].Delta.Role = "assistant"
							isFirstChunk = false
						}
						if err := yield(chunk); err != nil {
							return nil
						}
					}

					// Handle function calls
					if part.FunctionCall != nil {
						argsJSON, _ := json.Marshal(part.FunctionCall.Args)
						chunk := OpenAIChunk{
							ID:      id,
							Object:  "chat.completion.chunk",
							Created: created,
							Model:   p.model,
							Choices: []Choice{
								{
									Index: 0,
									Delta: Delta{
										ToolCalls: []ToolCall{
											{
												Index: toolCallIndex,
												ID:    fmt.Sprintf("call_%d", toolCallIndex),
												Type:  "function",
												Function: FunctionCall{
													Name:      part.FunctionCall.Name,
													Arguments: string(argsJSON),
												},
											},
										},
									},
									FinishReason: finishReason,
								},
							},
							Usage: usage,
						}
						if isFirstChunk {
							chunk.Choices[0].Delta.Role = "assistant"
							isFirstChunk = false
						}
						if err := yield(chunk); err != nil {
							return nil
						}
						toolCallIndex++
					}

					// Note: Gemini's part.Thought is a bool flag, not content.
					// Reasoning content is not exposed in the current SDK version.
				}
			}
		}
	}

	return nil
}

// mapGeminiFinishReason converts Gemini finish reasons to OpenAI format
func mapGeminiFinishReason(reason genai.FinishReason) string {
	switch reason {
	case genai.FinishReasonStop:
		return "stop"
	case genai.FinishReasonMaxTokens:
		return "length"
	case genai.FinishReasonSafety, genai.FinishReasonRecitation, genai.FinishReasonBlocklist:
		return "content_filter"
	case genai.FinishReasonOther:
		return "stop"
	default:
		return "stop"
	}
}

func (p *GeminiProvider) wrapError(err error) error {
	if err == nil {
		return nil
	}
	var gerr genai.APIError
	if errors.As(err, &gerr) {
		return &APIError{
			Message:    gerr.Message,
			StatusCode: gerr.Code,
			Status:     gerr.Status,
		}
	}
	return &APIError{Message: err.Error()}
}
